src/Databases/DatabasesOverlay.cpp

```diff
@@ -0,0 +1,266 @@
+#include <Databases/DatabasesOverlay.h>
+
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterCreateQuery.h>
+#include <Common/typeid_cast.h>
+
+#include <Databases/DatabaseMemory.h>
+#include <Databases/DatabaseFileSystem.h>
+
+#include <Storages/IStorage_fwd.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int CANNOT_GET_CREATE_TABLE_QUERY;
+}
+
+DatabasesOverlay::DatabasesOverlay(const String & name_, ContextPtr context_)
+    : IDatabase(name_), WithContext(context_->getGlobalContext()), log(&Poco::Logger::get("DatabaseOverlay(" + name_ + ")"))
+{
+}
+
+DatabasesOverlay & DatabasesOverlay::registerNextDatabase(DatabasePtr database)
+{
+    databases.push_back(std::move(database));
+    return *this;
+}
+
+bool DatabasesOverlay::isTableExist(const String & table_name, ContextPtr context_) const
+{
+    for (const auto & db : databases)
+    {
+        if (db->isTableExist(table_name, context_))
+            return true;
+    }
+    return false;
+}
+
+StoragePtr DatabasesOverlay::tryGetTable(const String & table_name, ContextPtr context_) const
+{
+    StoragePtr result = nullptr;
+    for (const auto & db : databases)
+    {
+        result = db->tryGetTable(table_name, context_);
+        if (result)
+            break;
+    }
+    return result;
+}
+
+void DatabasesOverlay::createTable(ContextPtr context_, const String & table_name, const StoragePtr & table, const ASTPtr & query)
+{
+    for (auto & db : databases)
+    {
+        try
+        {
+            db->createTable(context_, table_name, table, query);
```

Comment:
let's add a method to `IDatabase` named `isReadOnly()`, which will return `false` by default, and `true` for `DatbaseFilesystem`. Then here there will be no need in try catch, we can rethrow an exception if `!isReadOnly`. 
Because otherwise the behaviour can be unexpected: a user will be able to create tables with the same name in each database, but will be able to read only from the first one.

Reply:
fixed, thanks!